### ISSUE TITLE
Unable to check whether a RichTextBlock is empty

### ISSUE THREAD

**Issue by BertrandBordage** [torchbox/wagtail#2336]
When you use `RichTextBlock` in a template – especially when it’s in a `StructBlock` – you sometimes need to do something like this:
```html+django
{% if value.description %}
  <div class="something">
    {{ value.description }}
  </div>
{% endif %}
```
However, the test always evaluates to `True` because `value.description` is a `RichText` object that only implements a `__str__` method. This `__str__` always returns at least `'<div class="rich-text"></div>'`.

The solution is to add a `__bool__` method like this:
```python
    def __bool__(self):
        return bool(self.source)
```
Another solution may consist in adding this `<div class="rich-text"></div>` only if the block is not empty.

**Comment by gasman** [comment]
@BertrandBordage That makes sense, thanks! I wonder if this also affects our ability to set `required=True`...

(The one thing I'm not sure about - can we rely on the rich text widget to reliably return an empty string, rather than `<p></p>`? I guess that's something we can deal with separately, though.)

**Comment by BertrandBordage** [comment]
From what I saw in the current release (1.4rc1), saving an empty `RichTextBlock` always returns an empty string, but I know I saw the `<p></p>` issue on `RichTextField` in the past. The solution would be to clean the field as I did in my projects, but we of course need to make this robust and handle similar cases like `<p><em></em></p>`, `<p> </p>`, `<p>[any space here, non-breaking, narrow non breaking, etc]</p>`.

**Comment by JoshBarr** [comment]
another awesome reason that HTML is a poor storage format for richtext.

**Comment by BertrandBordage** [comment]
@JoshBarr But if we store it in another format, we have to chose between these solutions:

- parse HTML in order to store data, and a renderer for displaying it
- directly store the non-HTML form (JSON), and render it in Python for the front office and JavaScript for the back office
- directly store the JSON, and render it in Python on the front and back office, using AJAX for the back office while typing...

All three solutions will introduce more issues than fix. And I don't even think that would be easier to parse. HTML is complex and horrible for a reason: layout also is. How would you convert to JSON something as simple as this: `<p><em>a</em> <strong><a href="/">b</a></p>`?

**Comment by gasman** [comment]
Fixed in #2729

**Comment by rollcat** [comment]
In case anyone's interested, we've worked around the `<p></p>` stuff using a custom template tag:

```python
from html.parser import HTMLParser
from django import template

register = template.Library()


class TextExtractor(HTMLParser):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.text = ""

    def handle_starttag(self, tag, attrs):
        return

    def handle_endtag(self, tag):
        return

    def handle_data(self, data):
        self.text += data

    @classmethod
    def extract_text(cls, data):
        self = cls()
        self.feed(data)
        self.close()
        return self.text


@register.filter
def dehtmlize(value):
    return TextExtractor.extract_text(value)
```

The `html` module is available only in Python 3. Use like this:

```
{% if value.content.source | dehtmlize %} ...
```
